### ISSUE TITLE
push into synaptic queue in parallel per block

### ISSUE THREAD

**Issue by moritzaugustin** [brian-team/brian2cuda#16]
evaluate alternatives and implement after discussion the best:
- on initilization phase calculate per pre neuron (in the resp., post neuron block): the number of outgoing synapsis with delay 0dt, 1dt, ... Nmax*dt AND enforce sorting of the blocked connectivity matrix s.t. the synapses are ordered after delays => employ the number above parallel threads to push into the SAME delay queue in parallel AND in parallel for the different delays
- old idea (should be contained in above one): make an assignment of threadid-delayid (maybe outer loop necessary) to push into queue[delayid] in parallel for each block having multiple threads vs. currently one active thread (id 0) in the actual push phase
- also cf. _current_ brian2 cpp_standalone multithreading for ideas

add to the queue also the postneuron id to allow better parallel synaptic code objects, c.f. issue #15

**Comment by moritzaugustin** [comment]
extending what we have discussed the optimal way is to combine the two parallelism ways (over synapses per delay and over delays) and have a straight 1:1 correspondence between threads and synapses
still required: 
[*] sorting of synapses to groups with the same delay, i.e., d0,d0,d0,d1,d1,d1,d1,d1,d1,d1,d2,d2,d3,d3,d3,d3,d3 etc 
[*] save the no of syn. per delay for each preneuron (precomputation at initiliazation)
[*] extend vector with given amount from precompuation above
[*] insert into vector at given index)

reason:
[*]parallelism over the synapses of each delay only (serializing over the delays) might be too slow due to to little no of synapses per delay especially if we have uniform-like delay distribution (remember we have already grouped according to postsynaptic cells implemented as cuda blocks)
[*]parallelism over delays only (serializing over the synapses of each delay) on the other hand is suboptimal  w.r.t. memory coalescing

**Comment by denisalevi** [comment]
Parallel push implementation is working with [this commit](https://github.com/brian-team/brian2cuda/commit/9d359983ee532ce4954cc1b5a6c27b5c538b4782). All feature tests are passing, including STDP now. A simple test of mine usage Subgroups also work now. 

Further testing will be done once we updated to brian2 2.0, where we can run the brian2 test suite.

There are still open optimizations, that can easily be done, but which I didn't have time for yet. I opened a new issue for that (#31).

Closing this one.
